File: WarlockTools_src/WarlockMd.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;

namespace WarlockTools
{
    /// <summary>
    /// Warlock / Warlock 2 locdata.md (mdmp). Same magic as Majesty, different tables per game.
    /// Pool is UTF-8. W1 stamp 0xC4194; W2 stamp 0x101F70.
    /// Сделал Vlad302002, истинный Арданской король и покровитель Ардании. Слава Ардании!
    /// </summary>
    static class WarlockMd
    {
        public const uint Magic = 0x706D646D;

        sealed class Layout
        {
            public uint Stamp;
            public int DefaultCodepage;
            public int Block1Cap;
            public int BucketsMod;
            public int BucketsStore;
            public int MaxSlot;
            public int OvCap;
            public int PtrBase;
            public int HashHdr;
            public int Bucket0;
            public int BucketSize;
            public int OvOff;
            public int PoolUsedOff;
            public int PoolCapOff;
            public int PoolOff;
            public int PoolCap;
            public int Payload;
            public string Name;
        }

        // Warlock 1 — Master of the Arcane
        static readonly Layout W1 = new Layout
        {
            Name = "Warlock1",
            Stamp = 0xC4194,
            DefaultCodepage = 1251,
            Block1Cap = 13313,
            BucketsMod = 16641,
            BucketsStore = 16640,
            MaxSlot = 4,
            OvCap = 64,
            PtrBase = 0x34024,
            HashHdr = 0x34020,
            Bucket0 = 0x3402C,
            BucketSize = 56,
            OvOff = 0x117864,
            PoolUsedOff = 0x117B6C,
            PoolCapOff = 0x117B70,
            PoolOff = 0x117B74,
            PoolCap = 0xC0000,
            Payload = 0x1D7B80
        };

        // Warlock 2 — The Exiled
        static readonly Layout W2 = new Layout
        {
            Name = "Warlock2",
            Stamp = 0x101F70,
            DefaultCodepage = 1251,
            Block1Cap = 12288,
            BucketsMod = 8191,
            BucketsStore = 8191,
            MaxSlot = 7,
            OvCap = 100,
            PtrBase = 0x30014,
            HashHdr = 0x30010,
            Bucket0 = 0x3001C,
            BucketSize = 92,
            OvOff = 0xE7FC0,
            PoolUsedOff = 0xE8478,
            PoolCapOff = 0xE847C,
            PoolOff = 0xE8480,
            PoolCap = 0x100000,
            // 4 + payload = 2000020 (original l100 size; 16 B after pool region)
            Payload = 0x1E8490
        };

        static Layout LayoutByStamp(uint stamp)
        {
            if (stamp == W1.Stamp) return W1;
            if (stamp == W2.Stamp) return W2;
            throw new InvalidDataException("unexpected mdmp stamp 0x" + stamp.ToString("X") +
                " (known: W1=0xC4194, W2=0x101F70)");
        }

        static int GetKey(byte[] s, int bucketsMod)
        {
            int h = 0;
            for (int i = 0; i < s.Length; i++)
            {
                sbyte c = unchecked((sbyte)s[i]);
                h = h * 31 + c;
            }
            return (int)((uint)h % (uint)bucketsMod);
        }

        static int RdI(byte[] d, int o)
        {
            return BitConverter.ToInt32(d, o);
        }

        static uint RdU(byte[] d, int o)
        {
            return BitConverter.ToUInt32(d, o);
        }

        static void WrU(byte[] d, int o, uint v)
        {
            d[o] = (byte)v;
            d[o + 1] = (byte)(v >> 8);
            d[o + 2] = (byte)(v >> 16);
            d[o + 3] = (byte)(v >> 24);
        }

        static void WrI(byte[] d, int o, int v)
        {
            WrU(d, o, unchecked((uint)v));
        }

        static byte[] Resolve(byte[] d, Layout L, int index)
        {
            int slot = index / L.BucketsMod;
            int sl;
            if (slot < L.MaxSlot)
            {
                int buck = index % L.BucketsMod;
                sl = L.Bucket0 + buck * L.BucketSize + 8 + slot * 12;
            }
            else
                sl = L.OvOff + 8 + (slot - L.MaxSlot) * 12;
            int ptr = RdI(d, sl);
            int ln = RdI(d, sl + 4);
            int fo = ptr + L.PtrBase;
            if (ln < 0 || fo < 0 || fo + ln > d.Length)
                throw new InvalidDataException("bad string ptr");
            var s = new byte[ln];
            Buffer.BlockCopy(d, fo, s, 0, ln);
            return s;
        }

        public static List<KeyValuePair<string, string>> Unpack(byte[] data)
        {
            if (data.Length < 16 || RdU(data, 0) != Magic)
                throw new InvalidDataException("not mdmp");
            Layout L = LayoutByStamp(RdU(data, 4));
            int n = RdI(data, 0x0C);
            if (n < 0 || n > L.Block1Cap)
                throw new InvalidDataException("pairCount");
            var list = new List<KeyValuePair<string, string>>(n);
            var enc = new UTF8Encoding(false, false);
            for (int i = 0; i < n; i++)
            {
                int o = 0x10 + i * 16;
                string name = enc.GetString(Resolve(data, L, RdI(data, o + 4)));
                string val = enc.GetString(Resolve(data, L, RdI(data, o + 12)));
                list.Add(new KeyValuePair<string, string>(name, val));
            }
            return list;
        }

        public static void UnpackFile(string mdPath, string txtPath)
        {
            byte[] data = File.ReadAllBytes(mdPath);
            if (data.Length < 16 || RdU(data, 0) != Magic)
                throw new InvalidDataException("not mdmp");
            uint stamp = RdU(data, 4);
            Layout L = LayoutByStamp(stamp);
            int codepage = RdI(data, 8);
            var pairs = Unpack(data);
            var sb = new StringBuilder(pairs.Count * 40);
            sb.AppendFormat("{0:x8} {1:x8}\r\n", stamp, codepage);
            foreach (var p in pairs)
            {
                // Normalize CR/LF so bare \r cannot split lines on re-pack.
                string v = p.Value.Replace("\r\n", "\n").Replace("\r", "\n").Replace("\n", "\\n");
                sb.Append(p.Key);
                sb.Append('=');
                sb.Append(v);
                sb.Append("\r\n");
            }
            File.WriteAllText(txtPath, sb.ToString(), new UTF8Encoding(false));
        }

        sealed class Table
        {
            public readonly List<int[]>[] Buckets;
            public readonly List<int[]> Overflow = new List<int[]>();
            public readonly List<byte> Pool = new List<byte> { 0 };
            public readonly Dictionary<string, int> Intern = new Dictionary<string, int>(StringComparer.Ordinal);
            public readonly int Ptr0;
            readonly Layout L;

            public Table(Layout layout)
            {
                L = layout;
                Buckets = new List<int[]>[L.BucketsStore];
                for (int i = 0; i < L.BucketsStore; i++)
                    Buckets[i] = new List<int[]>();
                Ptr0 = L.PoolOff - L.PtrBase;
            }

            public int Add(byte[] s, int pairIdx)
            {
                string key = Encoding.UTF8.GetString(s);
                int enc;
                if (Intern.TryGetValue(key, out enc))
                    return enc;
                int ptr = Ptr0 + Pool.Count;
                Pool.AddRange(s);
                Pool.Add(0);
                int ln = s.Length;
                int buck = GetKey(s, L.BucketsMod);
                int ix = (s.Length > 0 && s[0] == (byte)'#') ? pairIdx : -1;
                var rec = new[] { ptr, ln, ix };
                int slot;
                if (buck < L.BucketsStore && Buckets[buck].Count < L.MaxSlot)
                {
                    slot = Buckets[buck].Count;
                    Buckets[buck].Add(rec);
                }
                else
                {
                    slot = L.MaxSlot + Overflow.Count;
                    Overflow.Add(rec);
                }
                enc = buck + slot * L.BucketsMod;
                Intern[key] = enc;
                return enc;
            }
        }

        static byte[] Pack(List<KeyValuePair<string, string>> pairs, Layout L, int codepage)
        {
            if (pairs.Count > L.Block1Cap)
                throw new InvalidDataException("too many pairs");
            var t = new Table(L);
            var b1 = new List<int[]>();
            var utf8 = new UTF8Encoding(false);
            for (int i = 0; i < pairs.Count; i++)
            {
                int ni = t.Add(utf8.GetBytes(pairs[i].Key), i);
                int vi = t.Add(utf8.GetBytes(pairs[i].Value), i);
                int name0 = i == 0 ? L.Block1Cap : 0;
                b1.Add(new[] { name0, ni, 0, vi });
            }
            if (t.Overflow.Count > L.OvCap)
                throw new InvalidDataException("hash overflow (>" + L.OvCap + ")");
            if (t.Pool.Count > L.PoolCap)
                throw new InvalidDataException("string pool too big");

            var buf = new byte[4 + L.Payload];
            WrU(buf, 0, Magic);
            WrU(buf, 4, L.Stamp);
            WrU(buf, 8, (uint)codepage);
            WrI(buf, 0x0C, pairs.Count);
            for (int i = 0; i < L.Block1Cap; i++)
            {
                int o = 0x10 + i * 16;
                if (i < b1.Count)
                {
                    WrI(buf, o, b1[i][0]);
                    WrI(buf, o + 4, b1[i][1]);
                    WrI(buf, o + 8, b1[i][2]);
                    WrI(buf, o + 12, b1[i][3]);
                }
                else
                {
                    WrI(buf, o, 0);
                    WrI(buf, o + 4, -1);
                    WrI(buf, o + 8, 0);
                    WrI(buf, o + 12, -1);
                }
            }
            WrU(buf, L.HashHdr, 0);
            WrU(buf, L.HashHdr + 4, (uint)L.BucketsMod);
            WrU(buf, L.HashHdr + 8, (uint)L.BucketsMod);
            for (int b = 0; b < L.BucketsStore; b++)
            {
                int o = L.Bucket0 + b * L.BucketSize;
                var slots = t.Buckets[b];
                WrI(buf, o, slots.Count);
                WrI(buf, o + 4, L.MaxSlot);
                for (int s = 0; s < L.MaxSlot; s++)
                {
                    int sl = o + 8 + s * 12;
                    if (s < slots.Count)
                    {
                        WrI(buf, sl, slots[s][0]);
                        WrI(buf, sl + 4, slots[s][1]);
                        WrI(buf, sl + 8, slots[s][2]);
                    }
                    else
                    {
                        WrI(buf, sl, 0);
                        WrI(buf, sl + 4, 0);
                        WrI(buf, sl + 8, -1);
                    }
                }
            }
            WrI(buf, L.OvOff, t.Overflow.Count);
            WrI(buf, L.OvOff + 4, L.OvCap);
            for (int s = 0; s < L.OvCap; s++)
            {
                int sl = L.OvOff + 8 + s * 12;
                if (s < t.Overflow.Count)
                {
                    WrI(buf, sl, t.Overflow[s][0]);
                    WrI(buf, sl + 4, t.Overflow[s][1]);
                    WrI(buf, sl + 8, t.Overflow[s][2]);
                }
                else
                {
                    WrI(buf, sl, 0);
                    WrI(buf, sl + 4, 0);
                    WrI(buf, sl + 8, -1);
                }
            }
            WrI(buf, L.PoolUsedOff, t.Pool.Count);
            WrI(buf, L.PoolCapOff, L.PoolCap);
            t.Pool.CopyTo(buf, L.PoolOff);
            // W2 originals end with a tiny locale tag after the pool region.
            if (L.Stamp == W2.Stamp)
            {
                int fo = L.PoolOff + L.PoolCap; // 0x1E8480
                string tag = codepage == 1252 ? "en" : "ru";
                WrI(buf, fo, 2);
                WrI(buf, fo + 4, 8);
                byte[] tb = Encoding.ASCII.GetBytes(tag);
                Buffer.BlockCopy(tb, 0, buf, fo + 8, tb.Length);
            }
            return buf;
        }

        public static void PackFile(string txtPath, string mdPath)
        {
            var pairs = new List<KeyValuePair<string, string>>();
            Layout L = W1;
            int codepage = W1.DefaultCodepage;
            bool layoutSet = false;
            foreach (string raw in File.ReadAllLines(txtPath, Encoding.UTF8))
            {
                string line = raw.TrimEnd('\r');
                if (line.Length == 0) continue;
                if (!layoutSet && line[0] != '#' && line.IndexOf('=') < 0)
                {
                    // header: "000c4194 000004e3" or "00101f70 000004e4"
                    string[] parts = line.Split((char[])null, StringSplitOptions.RemoveEmptyEntries);
                    if (parts.Length >= 1)
                    {
                        uint stamp;
                        if (uint.TryParse(parts[0], NumberStyles.HexNumber, CultureInfo.InvariantCulture, out stamp))
                        {
                            L = LayoutByStamp(stamp);
                            layoutSet = true;
                        }
                    }
                    if (parts.Length >= 2)
                    {
                        int cp;
                        if (int.TryParse(parts[1], NumberStyles.HexNumber, CultureInfo.InvariantCulture, out cp))
                            codepage = cp;
                    }
                    continue;
                }
                if (line[0] != '#' && line.IndexOf('=') < 0) continue;
                int eq = line.IndexOf('=');
                if (eq <= 0) continue;
                string k = line.Substring(0, eq);
                string v = line.Substring(eq + 1).Replace("\\n", "\n");
                pairs.Add(new KeyValuePair<string, string>(k, v));
            }
            File.WriteAllBytes(mdPath, Pack(pairs, L, codepage));
        }
    }
}

File: WarlockTools_src/warlock_md.py
# Warlock locdata.md (mdmp) — not Majesty layout.
# UTF-8 string pool, 16641-mod hash, 4 slots/bucket, overflow list.

import struct
import hashlib
from pathlib import Path

MAGIC = 0x706D646D
STAMP = 0xC4194
CODEPAGE = 1251
BLOCK1_CAP = 13313
BUCKETS_MOD = 16641
BUCKETS_STORE = 16640
MAX_SLOT = 4
OV_CAP = 64
PTR_BASE = 0x34024
HASH_HDR = 0x34020
BUCKET0 = 0x3402C
OV_OFF = 0x117864
POOL_USED_OFF = 0x117B6C
POOL_CAP_OFF = 0x117B70
POOL_OFF = 0x117B74
POOL_CAP = 0xC0000
PAYLOAD = 0x1D7B80  # bytes after magic the exe reads


def _u32(d, o):
    return struct.unpack_from("<I", d, o)[0]


def _i32(d, o):
    return struct.unpack_from("<i", d, o)[0]


def get_key(s: bytes) -> int:
    h = 0
    for b in s:
        c = b if b < 128 else b - 256
        h = h * 31 + c
        h = ((h + 0x80000000) & 0xFFFFFFFF) - 0x80000000
    return (h & 0xFFFFFFFF) % BUCKETS_MOD


def _resolve(d, index):
    slot = index // BUCKETS_MOD
    buck = index % BUCKETS_MOD
    if slot < MAX_SLOT:
        off = BUCKET0 + buck * 56
        sl = off + 8 + slot * 12
        ptr, ln = _i32(d, sl), _i32(d, sl + 4)
    else:
        sl = OV_OFF + 8 + (slot - MAX_SLOT) * 12
        ptr, ln = _i32(d, sl), _i32(d, sl + 4)
    fo = ptr + PTR_BASE
    return d[fo : fo + ln]


def unpack_pairs(data: bytes):
    if _u32(data, 0) != MAGIC:
        raise ValueError("not mdmp")
    if _u32(data, 4) != STAMP:
        raise ValueError("unexpected header stamp 0x%X" % _u32(data, 4))
    n = _i32(data, 0x0C)
    pairs = []
    for i in range(n):
        o = 0x10 + i * 16
        name = _resolve(data, _i32(data, o + 4))
        val = _resolve(data, _i32(data, o + 12))
        pairs.append((name, val))
    return pairs


def unpack_to_text(md_path, txt_path):
    data = Path(md_path).read_bytes()
    pairs = unpack_pairs(data)
    lines = ["%08x %08x" % (STAMP, CODEPAGE)]
    for k, v in pairs:
        ks = k.decode("utf-8")
        vs = v.decode("utf-8").replace("\r\n", "\n").replace("\n", "\\n")
        lines.append("%s=%s" % (ks, vs))
    Path(txt_path).write_text("\n".join(lines) + "\n", encoding="utf-8")
    return len(pairs)


class _Table:
    def __init__(self):
        self.buckets = [[ ] for _ in range(BUCKETS_STORE)]  # list of (ptr, ln, idx)
        self.overflow = []
        self.pool = bytearray(b"\x00")  # empty string at 0
        self.intern = {}  # bytes -> encoded index
        self.ptr0 = POOL_OFF - PTR_BASE  # 0xE3B50

    def add(self, s: bytes, pair_idx: int) -> int:
        if s in self.intern:
            return self.intern[s]
        ptr = self.ptr0 + len(self.pool)
        self.pool += s + b"\x00"
        ln = len(s)
        buck = get_key(s)
        idx_field = pair_idx if s.startswith(b"#") else -1
        rec = (ptr, ln, idx_field)
        if buck < BUCKETS_STORE and len(self.buckets[buck]) < MAX_SLOT:
            slot = len(self.buckets[buck])
            self.buckets[buck].append(rec)
            enc = buck + slot * BUCKETS_MOD
        else:
            slot = MAX_SLOT + len(self.overflow)
            self.overflow.append(rec)
            enc = buck + slot * BUCKETS_MOD
        self.intern[s] = enc
        return enc


def pack_pairs(pairs):
    if len(pairs) > BLOCK1_CAP:
        raise ValueError("too many pairs: %d > %d" % (len(pairs), BLOCK1_CAP))
    t = _Table()
    block1 = []
    for i, (k, v) in enumerate(pairs):
        ni = t.add(k, i)
        vi = t.add(v, i)
        name_idx0 = BLOCK1_CAP if i == 0 else 0
        block1.append((name_idx0, ni, 0, vi))
    if len(t.overflow) > OV_CAP:
        raise ValueError("overflow overflow: %d > %d" % (len(t.overflow), OV_CAP))
    if len(t.pool) > POOL_CAP:
        raise ValueError("pool too big: %d > %d" % (len(t.pool), POOL_CAP))

    buf = bytearray(4 + PAYLOAD)
    struct.pack_into("<I", buf, 0, MAGIC)
    struct.pack_into("<I", buf, 4, STAMP)
    struct.pack_into("<I", buf, 8, CODEPAGE)
    struct.pack_into("<I", buf, 0x0C, len(pairs))
    for i in range(BLOCK1_CAP):
        o = 0x10 + i * 16
        if i < len(block1):
            a, b, c, d = block1[i]
        else:
            a, b, c, d = 0, -1, 0, -1
        struct.pack_into("<iiii", buf, o, a, b, c, d)
    struct.pack_into("<I", buf, HASH_HDR, 0)
    struct.pack_into("<I", buf, HASH_HDR + 4, BUCKETS_MOD)
    struct.pack_into("<I", buf, HASH_HDR + 8, BUCKETS_MOD)
    for b in range(BUCKETS_STORE):
        o = BUCKET0 + b * 56
        slots = t.buckets[b]
        struct.pack_into("<ii", buf, o, len(slots), MAX_SLOT)
        for s in range(MAX_SLOT):
            sl = o + 8 + s * 12
            if s < len(slots):
                ptr, ln, ix = slots[s]
                struct.pack_into("<iii", buf, sl, ptr, ln, ix)
            else:
                struct.pack_into("<iii", buf, sl, 0, 0, -1)
    struct.pack_into("<ii", buf, OV_OFF, len(t.overflow), OV_CAP)
    for s in range(OV_CAP):
        sl = OV_OFF + 8 + s * 12
        if s < len(t.overflow):
            ptr, ln, ix = t.overflow[s]
            struct.pack_into("<iii", buf, sl, ptr, ln, ix)
        else:
            struct.pack_into("<iii", buf, sl, 0, 0, -1)
    used = len(t.pool)
    struct.pack_into("<I", buf, POOL_USED_OFF, used)
    struct.pack_into("<I", buf, POOL_CAP_OFF, POOL_CAP)
    buf[POOL_OFF : POOL_OFF + used] = t.pool
    return bytes(buf)


def pack_from_text(txt_path, md_path):
    text = Path(txt_path).read_text(encoding="utf-8")
    pairs = []
    for line in text.splitlines():
        if not line.strip():
            continue
        if line[0] != "#" and " " in line[:20] and "=" not in line[:20]:
            continue  # header "000c4194 000004e3"
        if "=" not in line:
            continue
        k, v = line.split("=", 1)
        v = v.replace("\\n", "\n")
        pairs.append((k.encode("utf-8"), v.encode("utf-8")))
    data = pack_pairs(pairs)
    Path(md_path).write_bytes(data)
    return len(pairs), len(data)


if __name__ == "__main__":
    src = Path(r"E:\Steam\steamapps\common\Warlock - Master of the Arcane\modding.tools\_extract\l100\locdata.md")
    tmp = src.with_name("locdata_roundtrip.md")
    txt = src.with_name("locdata_dec.txt")
    n = unpack_to_text(src, txt)
    print("unpacked", n, "->", txt)
    n2, sz = pack_from_text(txt, tmp)
    print("packed", n2, "bytes", sz)
    a = src.read_bytes()
    b = tmp.read_bytes()
    print("orig", len(a), hashlib.md5(a).hexdigest())
    print("new ", len(b), hashlib.md5(b).hexdigest())
    if a[: len(b)] == b or a == b:
        print("IDENTICAL (or prefix)")
    else:
        diffs = 0
        m = min(len(a), len(b))
        first = None
        for i in range(m):
            if a[i] != b[i]:
                diffs += 1
                if first is None:
                    first = i
        print("diffs", diffs, "first", hex(first) if first is not None else None)
        if first is not None:
            print(" orig", a[first : first + 32].hex())
            print(" new ", b[first : first + 32].hex())
